core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -474,12 +470,23 @@ public boolean process()
                 return false;
             }
 
-            // putIfAbsent will rehash automatically if rehash is needed, unless there isn't enough memory to do so.
-            // Therefore needRehash will not generally return true even if we have just crossed the capacity boundary.
-            while (lastPosition < positionCount && !needRehash()) {
-                int positionInDictionary = block.getId(lastPosition);
-                registerGroupId(dictionary, positionInDictionary);
-                lastPosition++;
+            int remainingPositions = positionCount - lastPosition;
+
+            while (remainingPositions != 0) {
+                int positionCountUntilRehash = maxFill - nextGroupId;
+                if (positionCountUntilRehash == 0) {
```

Comment:
use
```
if (positionCountUntilRehash < BATCH_SIZE)
```
Some memory overhead is fine rather than some odd query that would perform worse because `positionCountUntilRehash` is `1`